lib/demuxeres/gstdemuxeres.c

```diff
@@ -0,0 +1,721 @@
+  }
+  //Look for stream_start or eos event
+  while ((object =
+          gst_app_sink_try_pull_object (GST_APP_SINK (priv->appsink), -1))) {
+    if (GST_IS_EVENT (object)) {
+      if (appsink_handle_event (demuxer, GST_EVENT (object))) {
+        break;
+      }
+    } else if (GST_IS_SAMPLE (object)) {
+      priv->pending_sample = GST_SAMPLE (object);
+      break;
+    }
+  }
+
+  return packet;
+}
+
+static GstDemuxerEStreamType
+gst_parse_stream_get_type_from_pad (GstPad * pad)
+{
+  GstCaps *caps;
+  const GstStructure *s;
+  const gchar *name;
+  GstDemuxerEStreamType type = DEMUXER_ES_STREAM_TYPE_UNKNOWN;
+
+  caps = gst_pad_query_caps (pad, NULL);
+
+  if (caps) {
+    s = gst_caps_get_structure (caps, 0);
+    name = gst_structure_get_name (s);
+    if (g_str_has_prefix (name, "video")) {
+      type = DEMUXER_ES_STREAM_TYPE_VIDEO;
+    } else if (g_str_has_prefix (name, "audio")) {
+      type = DEMUXER_ES_STREAM_TYPE_AUDIO;
+    } else if (g_str_has_prefix (name, "text")) {
+      type = DEMUXER_ES_STREAM_TYPE_TEXT;
+    }
+    gst_caps_unref (caps);
+  }
+  return type;
+}
+
+static GstDemuxerVideoCodec
+gst_parse_stream_get_vcodec_from_caps (GstCaps * caps)
+{
+  GstDemuxerVideoCodec ret = DEMUXER_ES_VIDEO_CODEC_UNKNOWN;
+  GstStructure *s;
+  const gchar *name;
+
+  if (!caps)
+    return ret;
+
+  s = gst_caps_get_structure (caps, 0);
+  if (!s)
+    goto beach;
+
+  name = gst_structure_get_name (s);
+  if (!name)
+    goto beach;
+
+  if (!strcmp (name, "video/x-h264")) {
+    ret = DEMUXER_ES_VIDEO_CODEC_H264;
+  } else if (!strcmp (name, "video/x-h265")) {
+    ret = DEMUXER_ES_VIDEO_CODEC_H265;
+  }
+
+beach:
+  return ret;
+}
+
+static GstDemuxerAudioCodec
+gst_parse_stream_get_acodec_from_caps (GstCaps * caps)
+{
+  GstDemuxerAudioCodec ret = DEMUXER_ES_AUDIO_CODEC_UNKNOWN;
+  GstStructure *s;
+  const gchar *name;
+
+  if (!caps)
+    return ret;
+
+  s = gst_caps_get_structure (caps, 0);
+  if (!s)
+    goto beach;
+
+  name = gst_structure_get_name (s);
+  if (!name)
+    goto beach;
+
+  if (!strcmp (name, "audio/x-aac")) {
+    ret = DEMUXER_ES_AUDIO_CODEC_AAC;
+  }
+
+beach:
+  return ret;
+}
+
+void
+gst_parse_stream_teardown (GstDemuxerEStream * stream)
+{
+  switch (stream->type) {
+    case DEMUXER_ES_STREAM_TYPE_VIDEO:
+      g_free (stream->data.video.profile);
+      g_free (stream->data.video.level);
+      break;
+    default:
+      break;
+  }
+  g_free (stream->stream_id);
+  g_free (stream);
+}
+
+GstDemuxerEStream *
+gst_parse_stream_create (GstDemuxerES * demuxer, GstPad * pad)
+{
+  GstDemuxerEStream *stream;
+  GstPad *funnel_pad;
+  GstDemuxerESPrivate *priv = demuxer->priv;
+  GstCaps *caps = gst_pad_query_caps (pad, NULL);
+
+
+  if (!caps) {
+    GST_ERROR
+        ("Unable to get the caps from pad, unable to create a new stream");
+    return NULL;
+  }
+  GST_INFO ("the stream caps is %" GST_PTR_FORMAT, caps);
+
+  stream = g_new0 (GstDemuxerEStream, 1);
```

Comment:
same as above with zeroing memory.

And I think this should be done at the GstStream event parsing.

Reply:
This g_new0 is called only once for each stream so I dont think its going to kill the performance and it allows to have the whole structure zero'ed.
When a new pad is created we have to plug it to the funnel, that's why I do all the job here.